Introduce `PartitionedChannel`

artembilan opened the pull request:
* Implement a `PartitionedChannel` as an extension of the `AbstractExecutorChannel`
* Supply this channel with a `PartitionedDispatcher` which is an extension of the `AbstractDispatcher`
* The target partition is essentially a `UnicastingDispatcher` with a single thead executor

<!--
Thanks for contributing to Spring Integration. 
Please provide a brief description of your pull-request and reference any related issue numbers (prefix references with #) or StackOverflow questions.

See the [Contributor Guidelines for more information](https://github.com/spring-projects/spring-integration/blob/main/CONTRIBUTING.adoc).
-->

artembilan committed:
Introduce `PartitionedChannel`

* Implement a `PartitionedChannel` as an extension of the `AbstractExecutorChannel`
* Supply this channel with a `PartitionedDispatcher` which is an extension of the `AbstractDispatcher`
* The target partition is essentially a `UnicastingDispatcher` with a single thead executor

artembilan commented:
If this is OK, I'll add some docs and DSL configuration.
Thanks

garyrussell commented on spring-integration-core/src/test/java/org/springframework/integration/channel/PartitionedChannelTests.java:
```diff
@@ -0,0 +1,108 @@
+					assertThat(messagesInPartition.get(0).getHeaders().get("partitionKey"))
+							.isEqualTo(messagesInPartition.get(1).getHeaders().get("partitionKey"));
+				});
+
```
If we match the thread name to the partition, we could assert that the records went to the right partition.

artembilan replied:
Well, it is not always true, since we don't use a partition key as is, but rather its `hashCode()`.
So, while all the same keys goes to the same partition, it does not mean that `1` for key will always determine the partition `0`.

garyrussell reviewed (changes requested):
Some suggestions; otherwise LGTM.

artembilan committed:
* Fix language in Javadocs
* Add docs for `PartitionedChannel` into `channel.adoc`
* Pre-populate partitions in the `PartitionedDispatcher`
to ensure a thread number reflection of the partition it is used for (for default `ThreadFactory`)
* Add Javadocs to `public` methods
* Add Java DSL `PartitionedChannelSpec` and respective factory methods into `Channels`
* Use `IntegrationMessageHeaderAccessor.CORRELATION_ID` header as a default partition key

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/dispatcher/PartitionedDispatcher.java:
```diff
@@ -0,0 +1,191 @@
+		return partitionDispatcher.dispatch(message);
+	}
+
+	private void prePopulatedPartitionsIfAny() {
```
```suggestion
	private synchronized void prePopulatedPartitionsIfAny() {
```

garyrussell commented on src/reference/asciidoc/channel.adoc:
```diff
@@ -203,6 +203,30 @@ CAUTION: The sender can sometimes block.
+[[partitioned-channel]]
+===== `PartitionedChannel`
+
+Starting with version 6.1, a `PartitionedChannel` implementation is provided.
+This is an extension of `AbstractExecutorChannel` and represents a point-to-point dispatching logic where an actual consumption happens on a specific thread determined by the partition key evaluated from a message sent to this channel.
+This channel is similar to the `ExecutorChannel` mentioned above, but with a difference that messages with the same partition key are always handled in the same thread preserving ordering.
+It does not require an external `TaskExecutor`, but can be configured with a custom `ThreadFactory` (e.g. `Thread.ofVirtual().name("partition-", 0).factory()`).
+This factory then used to populate single-thread executors into `MessageDispatcher` delegates per partition.
+By default, the `IntegrationMessageHeaderAccessor.CORRELATION_ID` message header is used for partition key.
+This channel can be configured as a simple bean:
```
```suggestion
Starting with version 6.1, a `PartitionedChannel` implementation is provided.
This is an extension of `AbstractExecutorChannel` and represents point-to-point dispatching logic where the actual consumption is processed on a specific thread, determined by the partition key evaluated from a message sent to this channel.
This channel is similar to the `ExecutorChannel` mentioned above, but with the difference that messages with the same partition key are always handled in the same thread, preserving ordering.
It does not require an external `TaskExecutor`, but can be configured with a custom `ThreadFactory` (e.g. `Thread.ofVirtual().name("partition-", 0).factory()`).
This factory is used to populate single-thread executors into a `MessageDispatcher` delegate, per partition.
By default, the `IntegrationMessageHeaderAccessor.CORRELATION_ID` message header is used as the partition key.
This channel can be configured as a simple bean:
```

garyrussell commented on src/reference/asciidoc/channel.adoc:
```diff
@@ -203,6 +203,30 @@ CAUTION: The sender can sometimes block.
+----
+====
+
+The channel will have `3` partitions - dedicated threads; will use a `partitionKey` header to determine in which partition the message must be handled.
```
```suggestion
The channel will have `3` partitions - dedicated threads; will use the `partitionKey` header to determine in which partition the message must be handled.
```

garyrussell replied:
```suggestion
The channel will have `3` partitions - dedicated threads; will use the `partitionKey` header to determine in which partition the message will be handled.
```

garyrussell reviewed (changes requested):
Just doc polishing.

artembilan committed:
* Fix language in Javadocs
* Fix Checkstyle violations
* Mark `PartitionedDispatcher.prePopulatedPartitions()` with `synchronized`

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/dispatcher/PartitionedDispatcher.java:
```diff
@@ -145,13 +145,15 @@ public void shutdown() {
 	@Override
 	public boolean dispatch(Message<?> message) {
-		prePopulatedPartitionsIfAny();
+		if (this.partitions.isEmpty()) {
```
This won't work, thread 2 might proceed before thread 1 has populated all partition execs, and NPE on the dispatcher returned by the `get()`.

artembilan replied:
How about now?
See the latest commit.

artembilan committed:
* Populate partitions from `PartitionedDispatcher` ctor
and when a custom `ThreadFactory` is set

garyrussell commented on spring-integration-core/src/main/java/org/springframework/integration/dispatcher/PartitionedDispatcher.java:
```diff
@@ -0,0 +1,191 @@
+	public void setThreadFactory(ThreadFactory threadFactory) {
+		Assert.notNull(threadFactory, "'threadFactory' must not be null");
+		this.threadFactory = threadFactory;
+		populatedPartitions();
```
Need to shut down and clear existing executors.

Perhaps call `shutdown()` first?

artembilan replied:
Yeah... Nothing to shutdown I guess: those executors have nit been used yet.
But we indeed have to `this.executors.clear()` .

garyrussell replied:
We don't know that; they could call this after we started processing messages.

garyrussell replied:
And if not, it's a no-op; so why worry?

artembilan replied:
Doh... We cannon populate them from the ctor: some other options are required for delegate dispatchers.
So, coming back to the `synchronized` solution 😄

artembilan committed:
* Clear `executors` in `populatedPartitions()`

artembilan committed:
* Bring back `synchronized populatedPartitions()` and use it in the `dispatch()`
